hydra-node/src/Hydra/Ledger/Cardano.hs

```diff
@@ -141,22 +147,45 @@ instance Crypto crypto => ToJSONKey (Cardano.TxIn crypto) where
 instance Crypto crypto => ToJSON (Cardano.TxOut (MaryEra crypto)) where
   toJSON (Cardano.TxOut addr value) =
     object
-      -- TODO: Use ledger's instance, which is a base16-encoded
-      -- serialized address. We might want to use bech32
-      -- serialization in the end.
-      [ "address" .= addr
+      [ "address" .= serialiseAddressBech32
       , "value" .= valueToJson value
       ]
+   where
+    -- Serialise addresses in bech32 including the prefix as standardized:
+    -- https://github.com/cardano-foundation/CIPs/blob/master/CIP-0005/CIP-0005.md
+    serialiseAddressBech32 =
+      Bech32.encodeLenient prefix . Bech32.dataPartFromBytes $ Cardano.serialiseAddr addr
+
+    -- REVIEW(SN): The ledger's 'Addr' type is bigger than we need here and we
+    -- are forced to come up with a prefix for Byron "bootstrap" addresses,
+    -- although they should actually be serialised differently..and would not be
+    -- relevant for Hydra in the first place.
+    prefix = case addr of
+      (Cardano.Addr Cardano.Mainnet _ _) -> [Bech32.humanReadablePart|addr|]
+      (Cardano.Addr Cardano.Testnet _ _) -> [Bech32.humanReadablePart|addr_test|]
+      (Cardano.AddrBootstrap _) -> [Bech32.humanReadablePart|addr_boot|]
 
 instance Crypto crypto => FromJSON (Cardano.TxOut (MaryEra crypto)) where
   parseJSON = withObject "TxOut" $ \o -> do
-    address <- o .: "address"
+    address <- o .: "address" >>= deserialiseAddressBech32
     value <- o .: "value" >>= valueParseJson
     pure $ Cardano.TxOut address value
    where
     valueParseJson = withObject "Value" $ \o ->
       Cardano.Value <$> o .: "lovelace" <*> pure mempty
 
+    deserialiseAddressBech32 t =
+      case Bech32.decodeLenient t of
+        Left err -> fail $ "failed to decode bech32: " <> show err
+        Right (prefix, dataPart) ->
+          let mAddr = Bech32.dataPartToBytes dataPart >>= Cardano.deserialiseAddr
+           in case (Bech32.humanReadablePartToText prefix, mAddr) of
+                ("addr", Just addr@(Cardano.Addr Cardano.Mainnet _ _)) -> pure addr
+                ("addr_test", Just addr@(Cardano.Addr Cardano.Testnet _ _)) -> pure addr
+                ("addr_boot", Just addr@(Cardano.AddrBootstrap _)) -> pure addr
```

Comment:
Do we care about all theses prefixes now? Would it not be simpler to just fail in all cases but the `addr` one?

Reply:
Then the roundtrip tests do not pass as we use a generator which emits all kinds of addresses.